### PR TITLE
[V2] Add node APIs and clean up packaging

### PR DESCRIPTION
**Pull request by shifucun**
- Add node.py with properties, property_values and triples API.
- Add key.py with API KEY util functions.
- Deprecate Python 2 support.
- Re-organize pip packing files .

### CODE REVIEW

File: datacommons/setup.py [hunk 1]
```diff
@@ -23,15 +25,10 @@
 URL = 'https://github.com/datacommonsorg/api-python'
 EMAIL = 'support@datacommons.org'
 AUTHOR = 'datacommons.org'
-REQUIRES_PYTHON = '>=2.7'
+REQUIRES_PYTHON = '>=3.7'
```
Comment (juliawu) [inline datacommons/setup.py:28]: Do we know if anyone is still using python 2.7 and should be alerted about this change?

File: datacommons/node.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+""" API to request node information.
+"""
+
+from typing import List
+
+from datacommons.requests import _post
+from datacommons.utils import _get_direction
+
+
+def properties(nodes: List[str], out: bool = True):
```
Comment (juliawu) [inline datacommons/node.py:23]: Suggestion: include type hints for return values.

**Comment by juliawu** [comment]
For my own understanding, is the intention for `property()`, `property_values()`, and `triples()` to supercede the current `get_property_labels()`, `get_property_values()`, and `get_triples()` while maintaining the same specs?

**Comment by juliawu** [comment]
LGTM, leaving approval for @beets or @pradh who may have more detailed comments.

File: datacommons/setup.py (hunk 1 above)
Comment (shifucun) [reply datacommons/setup.py:28]: The change will be pushed as a new pip version. So users with older version can still use python 2.7

**Comment by shifucun** [comment]
> For my own understanding, is the intention for `property()`, `property_values()`, and `triples()` to supercede the current `get_property_labels()`, `get_property_values()`, and `get_triples()` while maintaining the same specs?

They will supercede the current APIs but the specs won't necessarily be maintained, especially for observation APIs coming later.

File: datacommons/node.py (hunk 2 above)
Comment (shifucun) [reply datacommons/node.py]: Done

File: datacommons/node.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+""" API to request node information.
+"""
+
+from typing import Dict, List
+
+from datacommons.requests import _post
+from datacommons.utils import _get_direction
+
+
+def properties(nodes: List[str], out: bool = True) -> Dict[str, List[str]]:
```
Comment (beets) [inline datacommons/node.py:23]: since it's a bool, perhaps `is_out` or `is_out_arc` for the second parameter

File: datacommons/key.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" API key related functions.
+"""
+
+import os
+
+# Environment variable for API key.
+_KEY_ENV = 'DC_API_KEY'
```
Comment (pradh) [inline datacommons/key.py:20]: Does the user need to set it?  Should we talk about it in README/docs?

File: datacommons/node.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    """Retrieves all the properties for a list of nodes.
+    Args:
+        nodes: Node DCIDs.
+        out: Whether this is out going property.
+    """
+    resp = _post(f'/v1/bulk/properties/{_get_direction(out)}', {
+        'nodes': nodes,
+    })
+    result = {}
+    for item in resp.get('data', []):
```
Comment (beets) [inline datacommons/node.py:33]: should we iterate through the argument list of nodes instead to ensure the resulting object has all requested properties?

File: datacommons/node.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+"""
+
+from typing import Dict, List
+
+from datacommons.requests import _post
+from datacommons.utils import _get_direction
+
+
+def properties(nodes: List[str], out: bool = True) -> Dict[str, List[str]]:
+    """Retrieves all the properties for a list of nodes.
```
Comment (beets) [inline datacommons/node.py:24]: I'm not as familiar with pypi - if the docstring gets converted into documentation, do we need to put a Returns here so users know what to expect?

File: datacommons/node.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    result = {}
+    for item in resp.get('data', []):
+        node, properties = item['node'], item.get('properties', [])
+        result[node] = properties
+    return result
+
+
+def property_values(nodes: List[str],
+                    property: str,
+                    out: bool = True) -> Dict[str, List[str]]:
```
Comment (beets) [inline datacommons/node.py:41]: same here wrt `out` (and below)

File: docs/release.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# Python API Release
```
Comment (pradh) [inline docs/release.md:1]: Thanks for adding this!

File: datacommons/node.py (hunk 6 above)
Comment (beets) [inline datacommons/node.py:24]: Might be worth mentioning that we are not returning values, only available property keys?

File: datacommons/node.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        result[node] = []
+        for v in values:
+            if 'dcid' in v:
+                result[node].append(v['dcid'])
+            else:
+                result[node].append(v['value'])
+    return result
+
+
+def triples(nodes: List[str],
```
Comment (pradh) [inline datacommons/node.py:64]: should use the convention of <verb>_<thing>?

File: datacommons/node.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    """Retrieves the triples for a node.
+    Args:
+        node: Node DCID.
+        out: Whether the returned property is out-going for the queried nodes.
+    """
+    resp = _post(f'/v1/bulk/triples/{_get_direction(out)}',
+                 data={'nodes': nodes})
+    result = {}
+    for item in resp.get('data', []):
+        node, triples = item['node'], item.get('triples', {})
```
Comment (pradh) [inline datacommons/node.py:75]: I assume 'node' is always there?

File: datacommons/node.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            else:
+                result[node].append(v['value'])
+    return result
+
+
+def triples(nodes: List[str],
+            out: bool = True) -> Dict[str, Dict[str, List[object]]]:
+    """Retrieves the triples for a node.
+    Args:
+        node: Node DCID.
```
Comment (pradh) [inline datacommons/node.py:68]: nodes.  List of DCIDs?

File: datacommons/node.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        out: Whether the returned property is out-going for the queried nodes.
+    """
```
Comment (pradh) [inline datacommons/node.py:70]: Document return value?

File: datacommons/node.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+def property_values(nodes: List[str],
+                    property: str,
+                    out: bool = True) -> Dict[str, List[str]]:
+    """Retrieves the property values for a list of nodes.
+    Args:
+        nodes: Node DCIDs.
+        property: The property label to query for.
+        out: Whether this is out going property.
+    """
```
Comment (pradh) [inline datacommons/node.py:47]: Document return value/

File: datacommons/node.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+from datacommons.requests import _post
+from datacommons.utils import _get_direction
+
+
+def properties(nodes: List[str], out: bool = True) -> Dict[str, List[str]]:
+    """Retrieves all the properties for a list of nodes.
+    Args:
+        nodes: Node DCIDs.
+        out: Whether this is out going property.
```
Comment (pradh) [inline datacommons/node.py:27]: nit: Whether to return out going properties?

File: datacommons/requests.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    if api_key:
+        headers['x-api-key'] = api_key
+    try:
+        resp = requests.post(url, json=data, headers=headers)
+        if resp.status_code != 200:
+            raise Exception(
+                f'{resp.status_code}: {resp.reason}\n{resp.json()["message"]}')
+        return resp.json()
+    except requests.exceptions.Timeout:
+        raise Exception('data request timeout')
```
Comment (beets) [inline datacommons/requests.py:39]:
```suggestion
        raise Exception('Data request timed out, please try again.')
```

File: datacommons/node.py (hunk 5 above)
Comment (beets) [reply datacommons/node.py:33]: should we also add a try/catch for key errors so we can add a friendly error message

File: datacommons/setup.py [hunk 16]
```diff
@@ -23,15 +25,10 @@
...
 VERSION = '1.4.3'
-
-REQUIRED = [
-    'six',
-]
-
+REQUIRED = ['six', 'requests']
```
Comment (beets) [inline datacommons/setup.py:30]: do we still need six if we're fully on py 3?

File: datacommons/test/node_test.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        response = datacommons.properties(["City"])
+        assert response == {
+            "City": ["name", "provenance", "subClassOf", "typeOf"]
+        }
+
+
+class TestPropertyValues(unittest.TestCase):
+
+    @mock.patch("datacommons.node._post")
+    def test_with_data(self, _post):
```
Comment (beets) [inline datacommons/test/node_test.py:50]: can we also add a test with multiple values for a property?

**Review by pradh — COMMENTED** [review]
Nice!

Main question about function name...

**Review by beets — APPROVED** [review]
thanks for updating the library!

File: datacommons/key.py (hunk 4 above)
Comment (shifucun) [reply datacommons/key.py:20]: Yes, the user will need to set it. We should add this back in docs.

[the remark this answers is not in the document]
Comment (shifucun) [reply datacommons/key.py:20]: Yes, should pin our notebooks with the current version so they don't break.

File: datacommons/node.py (hunk 3 above)
Comment (shifucun) [reply datacommons/node.py]: Done

File: datacommons/node.py (hunk 6 above)
Comment (shifucun) [reply datacommons/node.py:24]: I think these docs are only meant for code, not for external documentations. I will put a short blurb of the return values

File: datacommons/node.py (hunk 14 above)
Comment (shifucun) [reply datacommons/node.py]: Done

File: datacommons/node.py (hunk 6 above)
Comment (shifucun) [reply datacommons/node.py:24]: Done

File: datacommons/node.py (hunk 5 above)
Comment (shifucun) [reply datacommons/node.py:38]: V1 REST will ensure all input have entries in the output, so here should be safe.

For error, can just let it throw the exception from inside _post(); it has added meaningful error msg there.

File: datacommons/node.py (hunk 7 above)
Comment (shifucun) [reply datacommons/node.py]: Done

File: datacommons/node.py (hunk 11 above)
Comment (shifucun) [reply datacommons/node.py]: Done

File: datacommons/node.py (hunk 10 above)
Comment (shifucun) [reply datacommons/node.py:87]: Yes, V1 always return the input node in each entry

File: datacommons/requests.py (hunk 15 above)
Comment (shifucun) [reply datacommons/requests.py]: Done

File: datacommons/setup.py (hunk 16 above)
Comment (shifucun) [reply datacommons/setup.py:30]: Will remove six after the old code is deleted.

File: datacommons/test/node_test.py (hunk 17 above)
Comment (shifucun) [reply datacommons/test/node_test.py:50]: Done

File: datacommons/node.py (hunk 12 above)
Comment (shifucun) [reply datacommons/node.py:82]: Done

File: datacommons/node.py (hunk 13 above)
Comment (shifucun) [reply datacommons/node.py:55]: Done

**Review by shifucun — COMMENTED** [review]
thanks for review!

**Review by pradh — APPROVED** [review]

**Pull request merged**